hs/src/Reach/Eval/Core.hs

```diff
@@ -4124,14 +4130,44 @@ evalPrim p sargs =
     SLPrim_Contract_new -> do
       at <- withAt id
       let lab = "new Contract"
-      (ccv, opts) <-
+      let expectContractCode = \case
+            SLV_ContractCode _ cc -> return cc
+            ow -> expect_t ow $ Err_Expected "ContractCode or Reach.App"
+      (cc, opts) <-
         case args of
-          [x] -> return (x, mempty)
-          [x, y] -> (,) x <$> mustBeObject y
+          [x@(SLV_Prim (SLPrim_App_Delay _ _ _ _ mname))] -> do
+            compileProg <- asks e_compileProg
+            name <- maybe (("Reach.App" <>) . show <$> ctxt_alloc) return mname
+            co <- compileProg name x
```

Comment:
I understand comment 2. I'm not sure I understand comment 3, do you mean something like this:

```
const name = case mname of 
  Just n -> originalWhich <> n 
  Nothing -> originalWhich <> ctr
```
?

Reply:
I guess what I mean is that I want to see something like

"Compiling coolKid for main" when there's a name, but the files are still named `${file}.coolKid.mjs`

But when there's no name it would be

"Compiled internal contract #2 for main" and the file is `${file}.${originalWhich}${counter}.mjs`

Reply:
sounds good